python/src/aiconfig/eval/metrics.py

```diff
@@ -98,52 +87,251 @@ def __lt__(self, other: Any) -> bool:
+                    raise ValueError(f"Error getting structured response: {e}")
+                case Ok(data):
+                    return common.CustomMetricPydanticObject(data=data)
+
+        return _evaluation_fn
+
+    return Metric(
+        evaluation_fn=_make_evaluation_fn(pydantic_basemodel_type),
+        metric_metadata=EvaluationMetricMetadata(
+            name=metric_name,
+            description=metric_description,
+            extra_metadata=dict(
+                basemodel_type_name=pydantic_basemodel_type.__name__,
+                eval_llm_name=eval_llm_name,
+                field_descriptions_json=json.dumps(field_descriptions, sort_keys=True),
+            ),
+        ),
+    )
+
+
+def _make_openai_structured_llm_metric_helper(
+    eval_llm_name: str,
+    pydantic_basemodel_type: Type[T_BaseModel],
+    metric_name: str,
+    metric_description: str,
+    field_descriptions: dict[str, str],
+    openai_chat_completion_create: OpenAIChatCompletionCreate | None = None,
+) -> Result[Metric[str, common.CustomMetricPydanticObject[T_BaseModel]], str]:
+    schema = pydantic_basemodel_type.model_json_schema()
+    properties = schema["properties"]
+    required = schema["required"]
+
+    if not field_descriptions.keys() <= properties.keys():
+        return Err(
+            f"""
+                The following field_descriptions keys are not in the schema:
+                {set(field_descriptions.keys()) - set(properties.keys())}
+            """
+        )
+
+    def _with_description(key: str, value: dict[str, str]) -> dict[str, str]:
+        if key in field_descriptions:
+            return cu.dict_union_allow_replace(value, {"description": field_descriptions[key]})
+        return value
+
+    properties = {k: _with_description(k, v) for k, v in properties.items()}
+
+    required = required or list(properties.keys())
+
+    openai_eval_llm_chat_completion_create: common.CompletionTextToSerializedJSON = make_fn_completion_text_to_serialized_json(
+        eval_llm_name=eval_llm_name,
+        properties=properties,
+        required=required,
+        openai_chat_completion_create=(openai_chat_completion_create or default_openai_chat_completion_create),
+    )
+
+    return Ok(
+        make_structured_llm_metric(
+            openai_eval_llm_chat_completion_create,
+            eval_llm_name=eval_llm_name,
+            pydantic_basemodel_type=pydantic_basemodel_type,
+            metric_name=metric_name,
+            metric_description=metric_description,
+            field_descriptions=field_descriptions,
+        )
+    )
+
+
+## User interface
+
+
+# 1. functions that return metrics intended to be called directly
+
+
+def make_openai_structured_llm_metric(
+    eval_llm_name: str,
+    pydantic_basemodel_type: Type[T_BaseModel],
+    metric_name: str,
+    metric_description: str,
+    field_descriptions: dict[str, str] = {},
+    openai_chat_completion_create: OpenAIChatCompletionCreate | None = None,
```

Comment:
What is this particular one for? Is this to enable [openai API introspection](https://aiconfig.lastmileai.dev/docs/overview/create-an-aiconfig/#openai-api-python-wrapper)? Curious why we wouldn't just go with the AIConfigRuntime interface instead

Reply:
You're right, ideally we should use aiconfig runtime here. Directly using OpenAI function calling was actually slightly easier / clearer to get going, so I did that in the interest of unblocking this. We should make it easier to get started with function calling using AIConfig. cc @Ankush-lastmile , @rossdanlm

Reply:
Also, `evaluate` uses asyncio.gather to run multiple blocking I/O calls concurrently, whereas AIConfig does not currently do that correctly (issue 434). We have to either fix that or pass batches back to the provider APIs to enable efficient inference and therefore eval.

Reply:
I also was asking why we are specifically using OpenAI params in https://github.com/lastmile-ai/aiconfig/pull/513#discussion_r1431914899

Reply:
hmm, as opposed to?